### PR TITLE
feat: List and List item

### PR DESCRIPTION
**Pull request by akash-codemonk**
## Description

List component along with ListItem.

Figma: https://www.figma.com/file/sz08OJt2VrEgxQ8eToFIqG/ADS2.0---Components?type=design&node-id=4948-17094&mode=design&t=WcnPFxY6V2lFNUvb-0
Notion doc: https://www.notion.so/appsmith/Action-List-5fbde9b45cdb43be9357d761f0c45dec

Fixes https://github.com/appsmithorg/appsmith/issues/27531

This isn't used anywhere yet in the main repo.

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] Manual on storybook 
- [ ] Manual on main repo
- [ ] Jest
- [ ] Cypress

### Issues raised during DP testing
https://github.com/appsmithorg/design-system/pull/649#issuecomment-1755124123

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: packages/design-system/src/__theme__/default/index.css [hunk 1]
```diff
@@ -155,8 +155,8 @@
   /* --ads-v2-color-outline: var(--ads-v2-color-blue-300);
   --ads-v2-border-width-outline: 2px; */
   /* TODO: fix focus issue across the platform */
-  --ads-v2-color-outline: transparent;
-  --ads-v2-border-width-outline: 0;
+  --ads-v2-color-outline: var(--ads-v2-color-blue-300);
```
Comment (albinAppsmith) [inline packages/design-system/src/__theme__/default/index.css:158]: Please revert this change. This was kept transparent intentionally.

File: packages/design-system/src/List/List.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+          </InlineDescriptionWrapper>
+        </ContentTextWrapper>
+      </StyledListItem>
+      {endIcon && (
+        <EndIconWrapper>
+          <Button
+            isDisabled={props.isDisabled}
+            isIconButton
+            kind="tertiary"
+            onClick={(e) => {
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.tsx:169]: Please add a handler function and avoid using anonymous function because it is not persistent when the component re-renders. This helps improving performance.

File: packages/design-system/src/List/List.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        data-disabled={props.isDisabled || false}
+        data-selected={props.selected}
+        endIcon={props.endIcon}
+        isBlockDescription={isBlockDescription}
+        onClick={props.onClick}
+        onKeyDown={listItemhandleKeyDown}
+        size={size}
+        tabIndex={props.isDisabled ? -1 : 0}
+      >
+        <ContentTextWrapper className="content-text-wrapper">
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.tsx:122]: Please follow ads class naming convention here.

**Comment by albinAppsmith** [comment]
@akash-codemonk  List item disabled story, item is still interactable.

**Comment by kavitasmoolya** [comment]
@akash-codemonk 

- [ ] List item size larger story  : Remove "items" control from here as it is already mentioned in "List" Story
image

**Comment by kavitasmoolya** [comment]
@akash-codemonk 

- [ ] Disable story : Item still looks like active and interactive
https://www.loom.com/share/4fa36fb3a259416f8af9b8855fccf7a5?sid=f3f1a1fe-6a83-45f5-a2cf-c3c80ce9e10d

- [ ] List story : I added additional item in the list but some how it is not reflecting in the list
https://www.loom.com/share/d858b89fc1264138a9144d4d8e956ed6?sid=d9c70e8d-33ac-46cd-9662-fe37ec925434

**Comment by akash-codemonk** [comment]
> * [ ]  List story : End icon is not visible for an item added in the list ("play-video") icon is not visible
>   end icon displays a shadow on hovering on it but icon is not visible
> * [ ]  List story : Few words lime "g" "y" Gets cutting off at the bottom if we add them in the description of any item in the list.

Unable to repro. Tried on windows, chrome.

list item

The icon is slightly cropped off because the size of the icon is too big for the button, other icons would work fine. 

> * [ ]  List story : Add long description of 3-4 lines and on hovering on it, tool tip gets overlapped with the title of the item

This is expected, the other option is to show it below which just seems off when there is enough space above.

> * [ ]  List story : End icon is not aligned with the title

End icon will be center aligned. 

@kavitasmoolya

File: packages/design-system/src/List/List.styles.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  height: 100%;
+  overflow: auto;
+  padding: var(--ads-v2-spaces-1);
+`;
+
+export const Wrapper = styled.div`
+  display: flex;
+  width: 100%;
+  align-items: center;
+  gap: 8px;
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.styles.tsx:40]: Please make use of spaces token like you did in other places.

File: packages/design-system/src/List/List.styles.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  ${(props) => !!props.endIcon && `padding: 8px 40px 8px 8px;`}}
+
+  ${({ size }) => Sizes[size]}
+
+  ${({ isBlockDescription, size }) =>
+    size === "lg" &&
+    isBlockDescription &&
+    `
+    ${DescriptionWrapper} {
+      padding-top: 3px
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.styles.tsx:111]: is  this value given in design? ideally we follow scale of 4px.

File: packages/design-system/src/List/List.styles.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    font-size: var(--listitem-bdescription-font-size);
+    line-height: normal;
+  }
+
+  & .${ListItemIDescClassName} {
+    font-size: var(--listitem-idescription-font-size);
+  }
+
+  &:hover {
+    background-color: var(--ads-v2-color-bg-subtle);
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.styles.tsx:140]: use `--ads-v2-colors-content-surface-hover-bg`

File: packages/design-system/src/List/List.styles.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  &:active {
+    background-color: var(--ads-v2-color-bg-muted);
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.styles.tsx:144]: use `--ads-v2-colors-content-surface-active-bg`

we should be using category tokens and not sdemantic

File: packages/design-system/src/List/List.styles.tsx [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  &[data-selected="true"] {
+    background-color: var(--ads-v2-color-bg-muted);
```
Comment (albinAppsmith) [inline packages/design-system/src/List/List.styles.tsx:148]: use `--ads-v2-colors-content-surface-active-bg`

**Review by albinAppsmith — APPROVED** [review]
